src/Open3D/Visualization/Rendering/Filament/FilamentResourceManager.cpp

```diff
@@ -0,0 +1,718 @@
+#include <filament/Scene.h>
+#include <filament/Skybox.h>
+#include <filament/Texture.h>
+#include <filament/TextureSampler.h>
+#include <filament/image/KtxBundle.h>
+#include <filament/image/KtxUtility.h>
+
+#include "FilamentEntitiesMods.h"
+#include "Open3D/GUI/Application.h"
+#include "Open3D/IO/ClassIO/ImageIO.h"
+#include "Open3D/Utility/Console.h"
+#include "Open3D/Utility/FileSystem.h"
+
+namespace open3d {
+namespace visualization {
+
+namespace {
+template <class ResourceType>
+using ResourcesContainer =
+        std::unordered_map<REHandle_abstract, std::shared_ptr<ResourceType>>;
+
+// We need custom shared pointer make function to
+// use engine deleter for allocated filament entities
+template <class ResourceType>
+std::shared_ptr<ResourceType> MakeShared(ResourceType* pointer,
+                                         filament::Engine& engine) {
+    return std::move(std::shared_ptr<ResourceType>(
+            pointer, [&engine](ResourceType* p) { engine.destroy(p); }));
+}
+
+template <class Handle, class ResourceType>
+Handle RegisterResource(filament::Engine& engine,
+                        ResourceType* resource,
+                        ResourcesContainer<ResourceType>& container) {
+    if (!resource) {
+        utility::LogError("Trying to register empty resource!");
+        return Handle::kBad;
+    }
+
+    auto newHandle = Handle::Next();
+    container[newHandle] = std::move(MakeShared(resource, engine));
+
+    return newHandle;
+}
+
+template <class ResourceType>
+std::weak_ptr<ResourceType> FindResource(
+        const REHandle_abstract& id,
+        ResourcesContainer<ResourceType>& container) {
+    auto found = container.find(id);
+    if (found != container.end()) {
+        return found->second;
+    }
+
+    utility::LogWarning("Resource {} not found.", id);
+    return std::weak_ptr<ResourceType>();
+}
+
+template <class ResourceType>
+void DestroyResource(const REHandle_abstract& id,
+                     ResourcesContainer<ResourceType>& container) {
+    auto found = container.find(id);
+    if (found == container.end()) {
+        utility::LogError("Trying to destroy nonexistent resource ({})!", id);
+        return;
+    }
+
+    container.erase(found);
+}
+
+// Image data that retained by renderer thread,
+// will be freed on PixelBufferDescriptor callback
+std::unordered_map<std::uint32_t, std::shared_ptr<geometry::Image>>
+        pendingImages;
+
+std::intptr_t RetainImageForLoading(
+        const std::shared_ptr<geometry::Image>& img) {
+    static std::intptr_t imgId = 1;
+
+    const auto id = imgId;
+    pendingImages[imgId] = img;
+    ++imgId;
+
+    return id;
+}
+
+void FreeRetainedImage(void* buffer, size_t size, void* userPtr) {
+    const auto id = reinterpret_cast<std::intptr_t>(userPtr);
+    auto found = pendingImages.find(id);
+    if (found != pendingImages.end()) {
+        pendingImages.erase(found);
+    } else {
+        utility::LogDebug(
```

Comment:
I'm not convinced we should die; we are freeing something that no longer exists, so whatever usage (if any) happened earlier. Going forward, whatever was using it no longer is. And since wee don't write to any images, there can't be any memory corruption (at least at present).
We need to have some sort of real solution for reporting programming errors, so I made a task in Asana.